File: harness/oracle/cases/context-compression.ts
/**
 * Oracle case #6: context-compression sizing (the pure subset).
 *
 * Drives the REAL pure functions from the v4 server's
 * lib/chat/context/compression.ts — the two compression-trigger predicates, the
 * sliding-window message split, and the compressed-history block builder. The
 * async LLM orchestrator (`applyContextCompression`) is deliberately NOT ported
 * here — it's a Phase-3 mocked-LLM target; only the side-effect-free sizing
 * logic is tier-1 material.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/context-compression.ts \
 *     > /tmp/oracle-context-compression.ndjson
 */

import {
  shouldApplyCompression,
  shouldApplyBudgetCompression,
  splitMessagesForCompression,
  buildCompressedHistoryBlock,
  type CompressibleMessage,
} from '@/lib/chat/context/compression';
import type { ContextCompressionSettings } from '@/lib/schemas/settings.types';

// The two predicates read only `enabled` (+ `windowSize` for the count trigger);
// build a partial settings object cast to the full type — faithful at runtime.
function settings(enabled: boolean, windowSize = 5): ContextCompressionSettings {
  return { enabled, windowSize } as unknown as ContextCompressionSettings;
}

type Msg = { role: 'user' | 'assistant' | 'system'; content: string };

type Row =
  | { kind: 'shouldApply'; id: string; messageCount: number; enabled: boolean; windowSize: number; bypass: boolean; out: boolean }
  | { kind: 'budget'; id: string; total: number; maxAvailable: number; enabled: boolean; bypass: boolean; out: boolean }
  | { kind: 'split'; id: string; messages: Msg[]; windowSize: number; out: { messagesToCompress: Msg[]; windowMessages: Msg[] } }
  | { kind: 'block'; id: string; input: string | null; out: string | null };

const rows: Row[] = [];

// ---------------------------------------------------------------------------
// shouldApplyCompression(messageCount, settings, bypass) — count trigger.
// ---------------------------------------------------------------------------
const shouldCases: Array<[string, number, boolean, number, boolean]> = [
  // [id, messageCount, enabled, windowSize, bypass]
  ['disabled', 10, false, 5, false],
  ['bypass', 10, true, 5, true],
  ['under-window', 3, true, 5, false],
  ['at-window', 5, true, 5, false],
  ['over-window', 6, true, 5, false],
  ['disabled-over', 10, false, 5, false],
  ['bypass-over', 10, true, 5, true],
];
for (const [id, messageCount, enabled, windowSize, bypass] of shouldCases) {
  rows.push({
    kind: 'shouldApply', id, messageCount, enabled, windowSize, bypass,
    out: shouldApplyCompression(messageCount, settings(enabled, windowSize), bypass),
  });
}

// ---------------------------------------------------------------------------
// shouldApplyBudgetCompression(total, maxAvailable, settings, bypass).
// ---------------------------------------------------------------------------
const budgetCases: Array<[string, number, number, boolean, boolean]> = [
  // [id, total, maxAvailable, enabled, bypass]
  ['disabled', 999, 1, false, false],
  ['bypass', 300, 100, true, true],
  ['under', 100, 200, true, false],
  ['equal', 200, 200, true, false],
  ['over', 201, 200, true, false],
];
for (const [id, total, maxAvailable, enabled, bypass] of budgetCases) {
  rows.push({
    kind: 'budget', id, total, maxAvailable, enabled, bypass,
    out: shouldApplyBudgetCompression(total, maxAvailable, settings(enabled), bypass),
  });
}

// ---------------------------------------------------------------------------
// splitMessagesForCompression(messages, windowSize) — sliding window split.
// ---------------------------------------------------------------------------
const msg = (i: number): Msg => ({ role: i % 2 === 0 ? 'user' : 'assistant', content: `m${i}` });
const seq = (n: number): Msg[] => Array.from({ length: n }, (_, i) => msg(i));
const splitCases: Array<[string, Msg[], number]> = [
  ['empty', seq(0), 5],
  ['under-window', seq(3), 5],
  ['at-window', seq(5), 5],
  ['over-window', seq(7), 5],
  ['window-1', seq(4), 1],
];
for (const [id, messages, windowSize] of splitCases) {
  rows.push({
    kind: 'split', id, messages, windowSize,
    out: splitMessagesForCompression(messages as CompressibleMessage[], windowSize),
  });
}

// ---------------------------------------------------------------------------
// buildCompressedHistoryBlock(compressedHistory) — empty/undefined → null.
// ---------------------------------------------------------------------------
const blockCases: Array<[string, string | undefined]> = [
  ['null', undefined],
  ['empty', ''], // falsy → null, same as undefined
  ['text', 'Alice met Bob at the fair.'],
];
for (const [id, input] of blockCases) {
  rows.push({
    kind: 'block', id, input: input ?? null,
    out: buildCompressedHistoryBlock(input),
  });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/context_compression.rs
//! Port of the pure sizing subset of lib/chat/context/compression.ts — the
//! sliding-window context-compression triggers, message split, and the
//! compressed-history block builder.
//!
//! Only the side-effect-free functions are ported here: the async LLM
//! orchestrator (`applyContextCompression`) is a Phase-3 mocked-LLM target, not
//! tier-1 material. These four decide WHETHER and WHAT to compress; the actual
//! cheap-LLM summarization lives above this seam.

/// The compression settings fields these functions read. (The full v4
/// `ContextCompressionSettings` carries token targets too, but the sizing
/// predicates only consult `enabled` and `windowSize`.)
#[derive(Clone, Copy, Debug)]
pub struct CompressionSettings {
    pub enabled: bool,
    pub window_size: i64,
}

/// A message in the conversation, simplified to what the split needs.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct CompressibleMessage {
    pub role: String,
    pub content: String,
}

/// Determines if compression should be applied based on message count and
/// settings — the legacy count-driven trigger. Disabled or bypassed short-circuit
/// to false; otherwise compress only when there are MORE messages than the window
/// (messages 1..=windowSize get full context).
pub fn should_apply_compression(
    message_count: i64,
    settings: &CompressionSettings,
    bypass_compression: bool,
) -> bool {
    if !settings.enabled {
        return false;
    }
    if bypass_compression {
        return false;
    }
    message_count > settings.window_size
}

/// Determines if budget-driven compression should be applied: when the total
/// estimated prompt tokens exceed the available budget. Disabled/bypass
/// short-circuit to false. (Only `enabled` is consulted from settings.)
pub fn should_apply_budget_compression(
    total_estimated_tokens: i64,
    max_available: i64,
    settings: &CompressionSettings,
    bypass_compression: bool,
) -> bool {
    if !settings.enabled {
        return false;
    }
    if bypass_compression {
        return false;
    }
    total_estimated_tokens > max_available
}

/// The result of a sliding-window split: the older messages to compress and the
/// recent ones to keep in full.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SplitResult {
    pub messages_to_compress: Vec<CompressibleMessage>,
    pub window_messages: Vec<CompressibleMessage>,
}

/// Split messages into those to compress and those to keep in full (the window).
///
/// When there are no more messages than the window, nothing is compressed and
/// every message is kept. Otherwise the last `window_size` messages are kept in
/// full and everything before them is compressed. (Assumes `window_size >= 1`,
/// which the v4 schema guarantees — `windowSize` is clamped to [3, 10]; the early
/// return makes the slice well-defined for every reachable input.)
pub fn split_messages_for_compression(
    messages: &[CompressibleMessage],
    window_size: i64,
) -> SplitResult {
    if (messages.len() as i64) <= window_size {
        // Nothing to compress — keep them all in the window.
        return SplitResult {
            messages_to_compress: Vec::new(),
            window_messages: messages.to_vec(),
        };
    }
    // Here window_size < messages.len(), so the split point is in range.
    let split = messages.len() - window_size as usize;
    SplitResult {
        messages_to_compress: messages[..split].to_vec(),
        window_messages: messages[split..].to_vec(),
    }
}

/// Build the compressed-history system block, or `None` when there's nothing to
/// emit. Mirrors the TS `!compressedHistory` falsiness: BOTH `None` and an empty
/// string yield `None` (an empty summary is not a block).
///
/// Emitted as a *separate* system message (the caller's concern) so the persona
/// prompt's cacheable bytes stay stable across turns — concatenating this on
/// would rehash the system prefix every time the rolling summary refreshed.
pub fn build_compressed_history_block(compressed_history: Option<&str>) -> Option<String> {
    match compressed_history {
        Some(h) if !h.is_empty() => Some(format!(
            "## Conversation Context (Compressed Summary of Earlier Messages)\n\n\
             The following is a summary of the earlier conversation. Recent messages follow this summary.\n\n\
             {h}"
        )),
        _ => None,
    }
}
